flax/training/checkpoints.py

```diff
@@ -94,16 +97,38 @@ def save_checkpoint(ckpt_dir,
   ckpt_tmp_path = _checkpoint_path(ckpt_dir, 'tmp', prefix)
   ckpt_path = _checkpoint_path(ckpt_dir, step, prefix)
   gfile.makedirs(os.path.dirname(ckpt_path))
+  base_path = os.path.join(ckpt_dir, prefix)
+  checkpoint_files = gfile.glob(base_path + '*')
+
+  if ckpt_path in checkpoint_files:
+    if not overwrite:
+      raise errors.InvalidCheckpointError(ckpt_path, step)
+  else:
+    checkpoint_files.append(ckpt_path)
+
+  checkpoint_files = natural_sort(checkpoint_files)
+  if ckpt_path != checkpoint_files[-1]:
+    if not overwrite:
+      raise errors.InvalidCheckpointError(ckpt_path, step)
+
   with gfile.GFile(ckpt_tmp_path, 'wb') as fp:
     fp.write(serialization.to_bytes(target))
 
   # Rename once serialization and writing finished.
-  gfile.rename(ckpt_tmp_path, ckpt_path)
+  gfile.rename(ckpt_tmp_path, ckpt_path, overwrite=overwrite)
   logging.info('Saved checkpoint at %s', ckpt_path)
+  print(ckpt_path)
+
+  # Remove newer checkpoints
+  if overwrite:
+    ind = checkpoint_files.index(ckpt_path) + 1
+    newer_ckpts = checkpoint_files[ind:]
+    checkpoint_files = checkpoint_files[:ind]
+    for path in newer_ckpts:
+      logging.info('Removing checkpoint at %s', path)
+      gfile.remove(path)
```

Comment:
probably fine since the default is not overwrite - but if people have this on a lot I'm sure someone will accidentally erase their checkpoints by accidentally re-using the same working directory. :)

Reply:
Actually you have to set overwrite=True and disable checkpoint restore.